From 09b09312b9847e025249b9ba5bb2071dd77a19c3 Mon Sep 17 00:00:00 2001
From: "ricea@chromium.org"
 <ricea@chromium.org@0039d316-1c4b-4281-b951-d872f2087c98>
Date: Wed, 29 Jan 2014 05:22:37 +0000
Subject: [PATCH] Fix UAF in WebSocketDispatcherHost

If Send() fails in SendOrDrop(), then |message| will have been deleted,
but SendOrDrop() will still try to access its header fields.

Fixed.

BUG=338354
TEST=content_unittests

Review URL: https://codereview.chromium.org/148263009

git-svn-id: svn://svn.chromium.org/chrome/trunk/src@247627 0039d316-1c4b-4281-b951-d872f2087c98
---
 content/browser/renderer_host/websocket_dispatcher_host.cc | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/content/browser/renderer_host/websocket_dispatcher_host.cc b/content/browser/renderer_host/websocket_dispatcher_host.cc
index 5ce07985d79eb5..ece71cbdf872b9 100644
--- a/content/browser/renderer_host/websocket_dispatcher_host.cc
+++ b/content/browser/renderer_host/websocket_dispatcher_host.cc
@@ -83,10 +83,13 @@ WebSocketHost* WebSocketDispatcherHost::GetHost(int routing_id) const {
 }
 
 WebSocketHostState WebSocketDispatcherHost::SendOrDrop(IPC::Message* message) {
+  const uint32 message_type = message->type();
+  const int32 message_routing_id = message->routing_id();
   if (!Send(message)) {
-    DVLOG(1) << "Sending of message type " << message->type()
+    message = NULL;
+    DVLOG(1) << "Sending of message type " << message_type
              << " failed. Dropping channel.";
-    DeleteWebSocketHost(message->routing_id());
+    DeleteWebSocketHost(message_routing_id);
     return WEBSOCKET_HOST_DELETED;
   }
   return WEBSOCKET_HOST_ALIVE;
